### PR TITLE
Two sided RSVD; matrix generation methods

### PR DESCRIPTION
**Pull request by chertila1**

### CODE REVIEW

File: tdecomp/matrix/decomposer.py [hunk 1]
```diff
@@ -167,24 +167,30 @@ def _lean_walsh_transform(
...
         half = s // 2
+        if half == 0:
+            break
         even = h[..., :half]
         odd = h[..., half:]
         h[..., :half] = even + odd
         h[..., half:] = even - odd
     
-    H = h.view_as(eye_k) * (1.0 / torch.sqrt(n))
+    H = h.view_as(eye_k) * (1.0 / torch.sqrt(torch.tensor(k, dtype=torch.float32)))
```
Comment (leostre) [inline tdecomp/matrix/decomposer.py:193]: функциональный торч спокойно кушает скаляры, можно не оборачивать в тензор

File: tensor-decompositions
Comment (leostre) [inline tensor-decompositions:1]: что это?

File: experiments/two_sided_run.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                torch.cuda.synchronize()
+            start_time = time.perf_counter()
+            
+            if method == torch.linalg.svd:
+                U, S, Vh = method(X, full_matrices=False)
+                X_approx = U @ torch.diag(S) @ Vh
+            else:
+                decomposer = method(**method_kwargs)
+                U, S, Vh = decomposer.decompose(X)
+                X_approx = decomposer._two_sided_compose(U, S, Vh)
```
Comment (leostre) [inline experiments/two_sided_run.py:88]: вот для такого можно использовать normalize_kwargs_universal

File: experiments/two_sided_run.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            for method in RANDOM_INIT_METHODS:
+                logger.info(f"Running {method} initialization...")
+                results = self.run_method(
+                    TwoSidedRandomSVD,
+                    matrix_gen,
+                    method_kwargs={"distortion_factor": 0.6, "random_init": method},
+                    n_repeats=N_REPEATS
+                )
+                all_results[method].append(self.aggregate_results(results))
+                self.clear_cuda_cache()
```
Comment (leostre) [inline experiments/two_sided_run.py:166]: ентот код и предыдущий дублируются, в следующий раз можно .run_method укрупнить встроив в него очистку кэша, агрегацию и логирование

File: tdecomp/matrix/decomposer.py [hunk 4]
```diff
@@ -248,27 +249,37 @@ class TwoSidedRandomSVD(Decomposer):
     }
 
     def __init__(self, *, distortion_factor: float = 0.6, 
-                 random_init: str = 'normal'):
+                 random_init: str = 'normal', rank: int):
```
Comment (leostre) [inline tdecomp/matrix/decomposer.py:252]: сначала указываются аргументы без дефолтного значения, потом с ним. 
+ rank можно сделать =None в таком случае ты моешь сделать if и вызывать get_stable_rank

**Review by leostre — APPROVED** [review]
